### ISSUE TITLE
Caffeine cache causing memory leak and OOM

### ISSUE THREAD

**Issue by mateusz-bajorek** [newrelic/newrelic-java-agent#558]
## Description
Recently, we've updated our services to Java 17 (v 17.0.1), which also forced us to use the NewRelic java agent version 7.4.0. 

After the migration, we've noticed our services regularly crash due to OOM. First, we've suspected some issue with Java 17, but after obtaining several heap dumps, it looks like it's a problem with Caffeine caching introduced in 7.3.0 #258 (we've migrated from an older version of the agent).

We believe it might be related to the MariaDB driver we're using. In one of the heapdumps, the `com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.WS` map has 2.1 mil entries of prepared statements. It could be the case as it would rely on `mariadb-java-client` implementation of `PreparedStatement` which does not implement `equals/hashCode`.

After doing two other heapdumps, one right after startup of the app and one an hour later, this is the comparison:

![image](https://user-images.githubusercontent.com/11185738/141983446-b0227822-a026-4d3d-b791-a8ec5cdeeacb.png)

Looks, like the cache contained only `PreparedStatement` entries.

## Expected Behavior
Agent and service running smoothly, no memory leak.

## Your Environment
NewRelic agent: `7.4.0`
Java: `eclipse-temurin:17` docker image (currently `17.0.1`)
Spring Boot: `2.5.6`
MariaDB driver: `2.7.4` ([managed by Spring Boot BOM](https://github.com/spring-projects/spring-boot/blob/v2.5.6/spring-boot-project/spring-boot-dependencies/build.gradle#L1077))

## Additional context
- [GTSE-12483](https://github.com/newrelic/newrelic-java-agent/issues/558)
- [GTSE-12474](https://newrelic.atlassian.net/jira/software/c/projects/GTSE/boards/1372?modal=detail&selectedIssue=GTSE-12474)

**Comment by Stephan202** [comment]
After upgrading several applications to Java 17 and New Relic Agent 7.3.0 (using a [fork](https://github.com/newrelic/newrelic-java-agent/issues/384#issuecomment-940219037) that includes Java 17 support), we also observed this issue with one of our applications. See the attached Eclipse Memory Analyzer screenshots:

<img src="https://user-images.githubusercontent.com/513280/142017389-f7341ba7-6923-4764-9dcf-77425685ba9b.png" width="45%" title="Leak suspects"> <img src="https://user-images.githubusercontent.com/513280/142017399-1fc94115-0ad7-4d2d-9bac-30e341b52b26.png" width="45%" title="Dominator tree">

The impacted application uses Oracle Java 17+35-2724, Spring Boot 2.5.5, PostgreSQL Driver 42.2.24, jOOQ 3.15.3 and HikariCP 3.4.5.

The screenshots show that many `PreparedStatement`s are retained by a Caffeine cache with weak keys. (I suspect it's [this cache](https://github.com/newrelic/newrelic-java-agent/blob/015220c26e6f27904ad282b7b1ade090d74e9917/newrelic-agent/src/main/java/com/newrelic/agent/instrumentation/weaver/extension/ExtensionHolderFactoryImpl.java#L35-L38), but don't yet understand the New Relic Agent code well enough to be sure.)

Inspecting some of the `WeakKeyReference`s that make up the `ConcurrentHashMap`'s keys, we notice that the `referent` field is (always?) `null`. This seems to indicate that the keys were garbage collected, which (based on the [documentation](https://github.com/ben-manes/caffeine/blob/f924b5c07c4defb148101cd1c17cc033df0acfb1/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Caffeine.java#L471-L494)) should have triggered the entries to be removed from the cache during [periodic maintenance](https://github.com/ben-manes/caffeine/blob/f924b5c07c4defb148101cd1c17cc033df0acfb1/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Caffeine.java#L117-L125). I wonder whether somehow this doesn't happen and whether manually calling `#cleanUp()` would help :thinking:. CC @ben-manes if this rings a bell; the agent [currently](https://github.com/newrelic/newrelic-java-agent/blob/56c6446c08134a3689666caf31c0477aecf9a66f/newrelic-agent/build.gradle#L96) bundles Caffeine 2.9.0.

**Comment by ben-manes** [comment]
Yes, that does seem strange. A write or a handful of reads should trigger a maintenance cycle which will remove these entries. If for some reason that is not being called (no activity) then `cleanUp()` would manually invoke the maintenance logic.

Internally the cache uses a `ReferenceQueue` so that the garbage collector will enqueue it when the referent is discarded. The maintenance logic will [poll this queue](https://github.com/ben-manes/caffeine/blob/f924b5c07c4defb148101cd1c17cc033df0acfb1/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1589-L1602). Any write or a few dozen reads will accumulate enough events in internal ring buffers to trigger a maintenance cycle to replay the activity on the policies and flush the reference queue. Unlike `Cleaner` we don't create dedicated threads blocked on this queue, so an idle cache won't be flushed automatically. You can of course use a `Cleaner` that calls `cleanUp` when the referent is null'd, but that is likely overkill. You could also call cleanUp on a scheduled basis. I am not aware of any bugs with that version which would break reference caching.

**Comment by ben-manes** [comment]
The drain status is used to avoid excessive locking and scheduling, so that only one maintenance task is running on the executor. The [scheduling](https://github.com/ben-manes/caffeine/blob/f924b5c07c4defb148101cd1c17cc033df0acfb1/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1505-L1528) sets the status to `PROCESSING_TO_IDLE`, a subsequent write will [transition](https://github.com/ben-manes/caffeine/blob/f924b5c07c4defb148101cd1c17cc033df0acfb1/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1492-L1496) it to `PROCESSING_TO_REQUIRED`, and the finally block in the maintenance will release it to `REQUIRED` before unlocking. If there was a race then it sounds like somehow this transition was dropped. Currently it is a plain write which is flushed by the lock's memory barrier. We could strengthen that if we think its the problem.

**Comment by Stephan202** [comment]
I see that on the 2.x branch [`Unsafe.putOrderedInt`](https://github.com/ben-manes/caffeine/blob/1afd872968c5a50cddb05a357e23fc21644a9a52/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L4007) is used instead of `VarHandle.setOpaque`, but IIUC that doesn't change the conclusion. On that branch too all `maintenance(..)` invocations are properly guarded by `evictionLock`. It's been many years since I read JCIP; not sure I'm qualified to provide suggestions here. :sweat_smile:

NB: Caffeine's 2.x line has seen more than 5 years of development and is widely deployed. Since the bug tracker doesn't list similar issues (iirc), I think it's rather interesting that this issue triggers in the context of a Java Agent. The fact that is a rather non-standard execution environment doesn't seem like a detail to be overlooked. (But I realize that this statement by itself isn't actionable.)

**Comment by ben-manes** [comment]
Is this reproducible and can we canary a patched version that uses a stronger memory barrier? We can use a volatile set and if that resolves it then I can cut a formal release.

**Comment by Stephan202** [comment]
In our case the issue happens frequently, so yes it's reproducible. :)

With a moderate amount of effort I can deploy a customized version of the agent, so yes, we can certainly run a test. Do you plan to provide a specific patch? I can take it from there. (Might take 1-3 days, though; I'm about to head off to bed, and my agenda for the coming days is very full.)

**Comment by ben-manes** [comment]
Yes. I'll provide a patch jar this evening (PST) and notify you here. If it helps then we can merge that in for a release. We can target the weekend to give us both some bandwidth if non-urgent. If urgent then I think a temporary workaround on your side is most prudent so that you can control the code.

**Comment by Stephan202** [comment]
Ah, a SNAPSHOT reference may even be easiest, as I can then directly wire it into our [fork](https://github.com/picnicSupermarket/newrelic-java-agent/) and reference/deploy it using Jitpack! (But just source is also fine; whatever is easiest for you.)

As always: thanks for the awesome community support. :pray:

**Comment by tbradellis** [comment]
Whoa, awesome! Yes, thank you @ben-manes @Stephan202 and @mateusz-bajorek  for this community effort.  We'll plan to update our dependency with the patch version in the next release.

**Comment by ben-manes** [comment]
As I debug the memory barriers, this is why I think the existing logic makes sense. That could be wrong, hardware bugs, OS bugs, JVM bugs, ARM vs x86's TSO, etc.

- Drain status starts at `IDLE` and may be CAS'd to `REQUIRED` on a write by `scheduleAfterWrite`.
- When `scheduleDrainBuffers` to schedule a maintenance task, under the lock it is lazySet to `PROCESSING_TO_IDLE` and the task submitted to an executor. The task submission or unlock will flush this value. At worst it could be overwritten by a CAS back to `REQUIRED`, which would have no harmful effects (maybe extra task if not started, else tryLock is a guard).
- The maintenance task completes by a `CAS(PROCESSING_TO_IDLE, IDLE)`. If that is successful then when unlocking a new task can be scheduled. A `CAS(IDLE, REQUIRED)` may happen before unlocking, which `scheduleDrainBuffers` ignores until the lock is available.
- If not successful, then a write in `scheduleAfterWrite` performed a `CAS(PROCESSING_TO_IDLE, PROCESSING_TO_REQUIRED)`. As only one write can win this CAS no other transitions are possible and won't be repeated. At this point only the maintenance task can set the drainStatus, which it does using as a `lazySet(REQUIRED)` on exit. That is flushed by the unlock, which makes it possible for further scheduling.
- An edge case scenario is when the write rate exceeds how fast we can drain the buffer. The `WRITE_BUFFER_MAX` threshold avoids excessively penalizing a thread and does a `lazySet(PROCESSING_TO_REQUIRED)`. In this scenario the writes would likely have coerced this and this is just an extra protection against leaving queued work by transitioning back to the `IDLE` state. That lazy set is by the maintenance thread so it has visibility reaching the `finally` block, where it uses a volatile read and program order should force its last write to be observed. Therefore this lazySet won't be flushed when later CAS'd to idle or set to required during the unlock step.

Because of this, I do not see a bug with the memory barriers to explain why it would be `PROCESSING_TO_REQUIRED` and unlocked. That can happen if the thread pool simply discards the task, as had happened in ForkJoinPool prior to u60 or with a `RejectedExecutionHandler`. In this case the cache can eventually recover when the writeBuffer fills up and `afterWrite` may block by calling `cleanUp()` to acquire the lock and perform the maintenance work directly, resetting the drain status. This can take a while (`128 * ceilingPowerOfTwo(NCPU)`), so you could accumulate dead entries until the cache can recover from this invalid state. If it was full and no drains happened, then you would see all writers blocked indefinitely on cache operations.

Does your heap dump show the write buffer's size? Do you think the executor might be silently discarding tasks?

---

While we discuss, you can try this [patch branch](https://github.com/ben-manes/caffeine/tree/v2.newrelic) using jitpack, `com.github.ben-manes:caffeine:v2.newrelic-SNAPSHOT`, or published into Sonatype's snapshot repository as `2.9.3-SNAPSHOT`.

**Comment by Stephan202** [comment]
(I did not (yet) compare the state transition/memory visibility analysis with the code, but on first read it makes sense to me too.) 

> Does your heap dump show the write buffer's size? 

I don't see a write buffer, which I think makes sense looking at the code: [`afterWrite`](https://github.com/ben-manes/caffeine/blob/1afd872968c5a50cddb05a357e23fc21644a9a52/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1363-L1389) guards the described logic by first checking `buffersWrites()`, which defaults to `false` and is not overridden by `WS`.

> Do you think the executor might be silently discarding tasks?

Since no custom executor is configured the cache uses the [default](https://github.com/ben-manes/caffeine/blob/1afd872968c5a50cddb05a357e23fc21644a9a52/caffeine/src/main/java/com/github/benmanes/caffeine/cache/Caffeine.java#L283-L310), i.e. `ForkJoinPool.commonPool()`. That executor [documents](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ForkJoinPool.html#execute(java.util.concurrent.ForkJoinTask)) it may throw `RejectedExecutionException`s. [This StackOverflow thread](https://stackoverflow.com/questions/69768534) suggests that `ForkJoinPool`'s rejection behavior changed between Java 16 and 17. (This may be relevant indeed: it's safe to say the vast majority of (New Relic/Caffeine) users has not yet upgraded to Java 17, while both reports in this issue _do_ reference Java 17. :thinking:) That said, I don't find [this warning message](https://github.com/ben-manes/caffeine/blob/1afd872968c5a50cddb05a357e23fc21644a9a52/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1437) in our logs. (Our own code logs at the debug level, but I assume that in this case logging is controlled by the New Relic agent, [which logs at the info level by default](https://docs.newrelic.com/docs/apm/agents/java-agent/configuration/java-agent-configuration-config-file/#cfg-log_level). In either case I don't expect the exception to be hidden.)

As for the Java 17 angle: @hughsimpson I see you added a thumbs-up to the OP. Do you also observe the issue, and if so: what does your environment look like?

> While we discuss [..]

Thanks! Will try to set aside some time to test this out :)

**Comment by ben-manes** [comment]
Oh, I forgot about `buffersWrites()` and being unnecessary here. That's a good catch and unfortunately means the self-correcting behavior is disabled because it is unnecessary for algorithmic reasons. The read buffer isn't needed here either, but is used as a cheap way to trigger the maintenance cycles.  It might be a good idea always enable the write buffer as a safety net from a misbehaving executor.

In a buggy JDK8, FJP had a race which caused it to [lose tasks](JDK-8078490) instead of running them. This would be equivalent to `ThreadPoolExecutor.DiscardPolicy`, which would break any state machine logic assumptions. This caused [evictions to halt](https://github.com/ben-manes/caffeine/issues/77) and a [memory leak](https://github.com/ben-manes/caffeine/issues/76) in Apache Druid, which is similar to your observations. In that case updating to a fixed JDK version solved it and prompted a switch to using a bounded write buffer. That failsafe bound allows us to have some recovery thanks to backpressure. 

Is there any chance your application still runs on JDK16? Did this by chance correspond with an upgrade?

**Comment by hughsimpson** [comment]
@Stephan202 Yep I'm seeing this same issue, also with with Java 17 and newrelic 7.4.0 - I'd wandered in here to report and saw I'd been pipped to the post by a few hours. Haven't tried with earlier versions of Java - I'm restricted to LTS versions, but could try with Java 11 and report back if I can find the bandwidth today

**Comment by mateusz-bajorek** [comment]
In our (ViaBill's) case, we had to disable NewRelic agent in our main app altogether, as the OOMs were too frequent, so we're a bit in the dark :) 

@ben-manes we observe this on Java 17 (using the `eclipse-temurin:17` docker image as the base).

Caffeine was introduced into the NR javaagent in version 7.3.0, compatibility with Java 17 was added in 7.4.0, so there is a chance it's related to the JDK used, as it's only one version apart.

**Comment by ben-manes** [comment]
Thanks, this is really interesting!

I think there are a few ways to test this theory,
a) Set `Caffeine.executor(Runnable::run)` as the executor for the java agent
b) Downgrade to an older JDK
c) Patch Caffeine to always buffer writes

I believe (c) is a good change regardless to safeguard against a discarding executor. That optimization is not worth the trouble you've all faced. If we can narrow down to a high degree of confidence that there is a FJP submission bug then we can escalate to Doug for advice on next steps.

**Comment by Stephan202** [comment]
1. Sure, I can update the fork and try with `Caffeine.executor(Runnable::run)`.
2. In our case the impacted application does use a Java 17 Docker image but still compiles to Java 11, so testing with a Java 16 runtime should be doable.

(I do have a bunch of meetings until end of day, so it might take a bit before I find a stretch of time long enough to test these things. Additionally the impacted application only sees load during the day, which might delay reproduction.)

**Comment by Stephan202** [comment]
I deployed [New Relic Agent 7.3.0 + Java 17 support + the custom Caffeine SNAPSHOT from Jitpack](https://github.com/PicnicSupermarket/newrelic-java-agent/compare/v7.3.0...128ba0d0ecbce5c928889672c2121dd230f6e382) to the affected application. It'll take a bit (possibly until tomorrow, since some key fulfillment center processes will soon wrap up) before I can say with certainty whether the issue still occurs. Will post another message once I know more.

**Comment by Stephan202** [comment]
Okay, I didn't wait for the application's memory usage to reach critical levels, but I can confirm that the patch doesn't make the issue go away. A new heap dump shows that `ExtensionHolderFactoryImpl.ExtensionHolderImpl.instanceCache` again references a large `WS` instance, now accounting for 47.58% of the heap. This time `keyReferenceQueue.queueLength` is 298808. (And `drainStatus` is again `PROCESSING_TO_REQUIRED`.)

I have now updated the New Relic Agent fork to include [one more commit](https://github.com/PicnicSupermarket/newrelic-java-agent/compare/v7.3.0...92919285371ed3cea8f45c034cc4c03681fc82cf), configuring `Caffeine.executor(Runnable::run)` for the relevant cache instance. I'll let this run over night and check again some time tomorrow morning. (The next two days are also full of meetings, so I won't be able to iterate much faster, sorry.)

**Comment by ben-manes** [comment]
Thanks @Stephan202! Then this is as expected that Caffeine's memory barrier wasn't the cause. This next test run should let us know if the executor is silently discarding the task.

**Comment by Stephan202** [comment]
Today the application used less memory, though there still was a slight upward trend. I created two heap dumps; in the mean time the application was restarted for unrelated reasons. Key findings:
- The `WS` instance referenced by `ExtensionHolderFactoryImpl.ExtensionHolderImpl.instanceCache` that contains the prepared statements is now small (just 267776 and 37752 bytes retained, respectively), with `keyReferenceQueue.queueLength == 0`, as it should be. :tada:

  All other `ExtensionHolderFactoryImpl.ExtensionHolderImpl.instanceCache` instances are now also small (on the order of kilobytes). I went back to yesterday's  heap dump and saw that a few other such instances were also larger back then (on the order of megabytes).

  All in all: whatever the issue/bug is, `Caffeine.executor(Runnable::run)` avoids it. :heavy_check_mark: 
- Two other `WS` instances that take up a significant amount of memory remain:
  - Instances referenced by `JdbcHelper.connectionToURL` comprising 22.17% / 8.12% of the heap, with a `keyReferenceQueue.queueLength` of 616446 / 102386.
  - Instances referenced by `JdbcHelper.connectionToIdentifier` comprising 18.47% / 5.74% of the heap, with a `keyReferenceQueue.queueLength` of 501032 / 76343.

  IIUC these caches are created by [`AgentCollectionFactory.createConcurrentWeakKeyedMap`](https://github.com/newrelic/newrelic-java-agent/blob/286968f7edef96aff1b072b50ea6dbc392afc024/newrelic-agent/src/main/java/com/newrelic/agent/util/AgentCollectionFactory.java#L18-L22), for which we did not set `Caffeine.executor(Runnable::run)`. It's safe to say that these two caches explain why memory usage was still trending upwards today.

As previously discussed, the next step would now be to test whether the issue is specific to Java 17, by downgrading to (e.g.) Java 16. Unfortunately my schedule today was even busier than expected, and tomorrow will certainly be no better. So instead I have started a different experiment that took me a bit less time to set up: I now [updated](https://github.com/PicnicSupermarket/newrelic-java-agent/compare/v7.3.0...a725fae71755e9aa6aae2189a06590c762ba73a7) the New Relic Agent to use `Executors.newCachedThreadPool()` rather than `Runnable::run` (this time also for the cache in `AgentCollectionFactory`). This experiment should hopefully tell us whether (a) the issue is somehow related to any kind of scheduling/thread switching or (b) it is more likely to be specific to `ForkJoinPool.commonPool()`.

Stay tuned :)

**Comment by ben-manes** [comment]
Wonderful! The openjdk github shows that changes were made to ForkJoinPool in jdk17, which had been untouched since jdk13.
- [JDK-8246585: ForkJoin updates](https://bugs.openjdk.java.net/browse/JDK-8246585) - [commit](https://github.com/openjdk/jdk/commit/5cfa8c94d6765b93f910fc01cd1ac2a56b16d78a)
- [JDK-8259800: timeout in tck test testForkJoin](https://bugs.openjdk.java.net/browse/JDK-8259800) - [commit](https://github.com/openjdk/jdk/commit/5b7b18c5bff575d2e94c5eadb14fcf1400e15d36)
- [JDK-8264572: ForkJoinPool.getCommonPoolParallelism() reports always 1](https://bugs.openjdk.java.net/browse/JDK-8264572) - [commit](https://github.com/openjdk/jdk/commit/cec66cf8efa469f3f8bf6af5da4bcf52b944c3e8)
- [JDK-8266459: Deprecate the Security Manager for Removal](https://bugs.openjdk.java.net/browse/JDK-8266459) - [commit](https://github.com/openjdk/jdk/commit/6765f902505fbdd02f25b599f942437cd805cad1#diff-e398beb49cd8d3e6c2f3a8ca8eee97172c57d7f88f3ccd8a3c704632cab32f5f)
- [JDK-8267521: maximum covering > 50K](https://bugs.openjdk.java.net/browse/JDK-8267521) - [commit](https://github.com/openjdk/jdk/commit/508cec7535cd0ad015d566389bc9e5f53ce4103b)

**Comment by Stephan202** [comment]
Alright, with `Executors.newCachedThreadPool()` the issue also does not reproduce. :tada: (The heap dump now contains only small `WS` instances, with the largest one (containing the prepared statements) retaining only 339046 bytes.) So this definitely points to the issue being somehow related to `ForkJoinPool.commonPool()`.

I have now repackaged the application based on Docker image [`openjdk:16.0.2-slim-bullseye`](https://hub.docker.com/layers/openjdk/library/openjdk/16.0.2-slim-bullseye/images/sha256-fe80319eac66c28760b26828b6b8ac373806b6d8a94d1fa75830b3bbd49ddd47) (Oracle Java 16.0.2+7-67) and deployed it in combination with the [original New Relic Agent fork](https://github.com/PicnicSupermarket/newrelic-java-agent/compare/v7.3.0...v7.3.0-picnic-1), i.e. without any of the modifications applied over the past few days. I hope to have a moment EOD to report any findings of this (hopefully final :upside_down_face:) experiment.

**Comment by Stephan202** [comment]
Alright, I can confirm that with Java 16 and without modifications the issue also doesn't reproduce, so the issue is restricted to Java 17 + `ForkJoinPool.commonPool()`. :heavy_check_mark:

(Grafana did indicate that overall memory usage now was higher and kept climbing much more than before, but this appears to be due to (a) the amount of committed memory going up over time, and (b) the size of the live set being consistently larger (but stable). Maybe the difference is due to some G1GC improvements in Java 17; not sure. In any case: with Java 16 the `WS` instances remain small, which is what we're after here.)

@ben-manes I have no experience with filing JDK issues, and you seem to have a firmer grasp on this stuff. So... what's next? :smile:

W.r.t. mitigation in the New Relic Agent, I see a two options:
1. Open a PR to set a custom executor on these caches. In this case we need to choose between `Runnable::run` and `Executors.newCachedThreadPool()`. The former is simpler, but the latter may yield better (startup?) performance. I didn't really check how/when these caches are used; @tbradellis and @jasonjkeller what's your feeling on this?
2. Wait a bit longer until we maybe know more, just in case there exists a nice Caffeine-side workaround. (This depends on what the actual root cause is, of course.) In that case we could upgrade to a hypothetical Caffeine 2.9.3 release.

On this topic: when is the next New Relic Agent release planned/how much time do we have to decide one way or another?

**Comment by ben-manes** [comment]
The next step is to loop in Doug Lea. I'll email him and we'll see what forum he prefers to discuss in (here, concurrency-interest, jira, ...). I tried to reproduce in a test case without caffeine and couldn't trigger the discarded task, so anything we can do to give him a reproducible error would be appreciated. You are welcome to email me if you'd like to be on that initial thread, as I lack everyone's contact details.

Almost always `Runnable::run` is acceptable and shouldn't impact performance. Since there are synchronous callbacks to user code, etc. it defaults to being asynchronous to minimize request latencies. The cache's own logic is efficient and O(1).

I will work on and release a patch that mitigates this issue by enabling `buffersWrites()` in all modes. This will allow the cache to recover when pending writes accumulate and it barges in to perform a maintenance cycle. That delay is not ideal, but enough to mask this problem and won't have much memory overhead.

The concern for JDK17 users is that if FJP discards tasks their own logic could be stuck. For example by default `CompletableFuture` uses that threadpool by default for async tasks. Since it is shared and you do not know who uses it and for what, it could cause other hard to reproduce instability problems.

**Comment by Stephan202** [comment]
> [..] so anything we can do to give him a reproducible error would be appreciated.

I'll give this some more thought. (I still wonder whether the fact that this happens in a Java agent is salient, but I have no idea to what extend Java agents are treated specially.)

> You are welcome to email me if you'd like to be on that initial thread, as I lack everyone's contact details.

My email address is my Github username + `@gmail.com`. (See also our correspondence April 21-24 earlier this year :).)

> I will work on and release a patch that mitigates this issue by enabling buffersWrites() in all modes.

Nice; I guess in that case I prefer option 2 above (i.e., just upgrading the dependency).

Thanks for all the help!

**Comment by ben-manes** [comment]
A quick status update.

I completed the work on v3, but have not yet backported it to v2 used here. Given the U.S. holiday week, I'll try but may not have a release until next weekend. Those changes only mask the problem because the ForkJoinPool may be in a corrupted state and impact the application's behavior.

@Stephan202 and Doug further isolated the problem to static initialization circularity. While there are some in progress improvements to relax this, this sounds like a mistake in NewRelic's agent which performs weaving during class loading. I suspect this is because Caffeine defaults to using `ForkJoinPool.commonPool()`, so constructing a cache forces that to initialize when it sets the internal executor field. This oversight might have occurred when replacing Guava's CacheBuilder usages with a Caffeine builder, as they appear to be nearly identical. However Guava lacks this option and simply runs work on a calling thread. That has the potential to degrade external latencies if user supplied logic is expensive, such as callbacks to a removal listener. Similar behavior can be obtained using `Runnable::run` which was shown to be fine for this workload.

A similar project is OpenTelemetry's java agent, which also embeds Caffeine for its weak keyed caches. They [switched](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2632) to a same-thread executor after other approaches to [wrangle](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3293/files) FJP become tiresome. /cc @anuraaga @trask

While preliminary findings, I believe the correct fix is to specify the `Caffeine.executor` in all instances created by the java agent.

**Comment by anuraaga** [comment]
Using the direct executor worked well for us in the javaagent, though we also have a hacky patch which I don't fully understand (maybe it wasn't needed thanks to the executor tweak).

https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/v1.8.0/instrumentation-api-caching/caffeine2/src/patch/java/io/opentelemetry/instrumentation/api/internal/shaded/caffeine2/cache

But for reference, we ended up having to swap Caffeine for CLHM because we realized instrumentation that caches may be used from Android - we only have some SQL caching that can affect instrumentation of SQLite. We may SPI in caffeine in the future but currently don't expect the difference between LFU and LRU to have a huge enough impact on this one cache.

**Comment by trask** [comment]
> @Stephan202 and Doug further isolated the problem to static initialization circularity

cool! I would love to see the repro / details if that's possible. always interested in understanding things to watch out for in java agent development

**Comment by ben-manes** [comment]
@trask currently we have not reproduced but reasoned through it based on heap dumps, Doug's deep knowledge, and code spelunking. The main bit of insight came from the email below,

**Comment by Stephan202** [comment]
> While preliminary findings, I believe the correct fix is to specify the `Caffeine.executor` in all instances created by the java agent.

Even if a future JDK release will avoid the issue, we should still do "something" for the JDK 17 releases already present in the wild. So I filed a PR based on your suggestion: #572.

Note that in the PR description I wrote "This may produce out-of-memory errors _as well as functional errors in the instrumented application_", because in the thread dumps I noticed the that FJP referenced not just Caffeine cache instances created by the New Relic agent, but also Caffeine instances created by our application. In other words: the broken FJP can also impact application code.

**Comment by tbradellis** [comment]
@Stephan202 

> On this topic: when is the next New Relic Agent release planned/how much time do we have to decide one way or another?

We have a release scheduled to come out in about 2 weeks.  I'm starting a branch that will use `Caffeine.executor(Runnable::run)` as @ben-manes suggested.  We'll have this ready to go in the next release and/or update the caffeine dependency as needed.

New twist. We have a customer reporting this in a support ticket on `IBM J9 VM JDK 8u301`.  The heap profile is the same.

**Comment by Stephan202** [comment]
>  I'm starting a branch that will use `Caffeine.executor(Runnable::run)` as @ben-manes suggested.

I see you already found #572 :). I noticed that you're no longer an outside collaborator on the fork, but if you'd like to have access again let me know and I'll send a new invite.

> We have a customer reporting this in a support ticket on `IBM J9 VM JDK 8u301`. The heap profile is the same.

That's interesting indeed. Not sure if we can see the source code of that JDK version, but the current JDK 17 implementation of `ForkJoinPool` uses `VarHandle`, so it's unlikely that those changes were backported all the way to Java 8. Maybe there's a subtly different root cause there. In any case, it can't hurt for them to try the `Caffeine.executor(Runnable::run)` changes as well, I suppose.

**Comment by tbradellis** [comment]
> I see you already found #572 :)
left some comments in the PR.

> Maybe there's a subtly different root cause there.
Makes sense. We'll of course continuing working with the customer and see if there's anything else we can surface around it.

**Comment by ben-manes** [comment]
If possible, you might want to initialize `ForkJoinPool.commonPool()` explicitly and early in the life cycle, in a manner that avoids circular initialization. There are some rare gotchas with j.u.c. and classloading for very core features, such as the [lost unpark bug](https://bugs.openjdk.java.net/browse/JDK-8074773). If you can safeguard during bootstrap then maybe you'll be protected from this happening again due to hidden usages that you did not know to account for. I'm not familiar enough with agents and weaving to know if that is possible and its implications, though.

**Comment by lsikora** [comment]
Hi. Sorry for bothering you in the comments, but I have trouble with finding a release date for this fix. I see it is labelled for 7.5.0 but when will this be released? I simply do not know if you wait for all targeted fixes to be applied or released at a given time. Would be awesome to get a source for this information. Thank you.

**Comment by Stephan202** [comment]
The observed issue where `ForkJoinPool.commonPool()` is not properly initialized turns out affect vanilla JDK 17 applications that have access to only a single core. See https://github.com/newrelic/newrelic-java-agent/pull/572#issuecomment-979441426 for my analysis. (In other words: the issue is not specific to the New Relic agent.)

@mateusz-bajorek @hughsimpson can you check whether your affected applications also run with only a single core? If so, consider explicitly setting `-Djava.util.concurrent.ForkJoinPool.common.parallelism=1` or assign an extra core. I'd be interested to hear whether that resolves the issue for you.

@lsikora do you also observe the issue? If so, let me know whether the above suggestions work for you. (As for the 7.5.0 release: https://github.com/newrelic/newrelic-java-agent/issues/558#issuecomment-975998134 indicates it's about two weeks out.)

**Comment by Stephan202** [comment]
One more cross-post (yeah: I should have kept things centrally in this thread): it turns out we're affected by a known JDK bug, fixed in the upcoming JDK 17.0.2. Details in https://github.com/newrelic/newrelic-java-agent/pull/572#issuecomment-979453789.

**Comment by hughsimpson** [comment]
@Stephan202 Had a sick kid this week so wasn't able to engage with this issue as I'd hoped. We're running in k8s so it's almost certainly https://bugs.openjdk.java.net/browse/JDK-8274349 as you suggest. Testing whether that arg resolves the leak now

**Comment by hughsimpson** [comment]
I will add that we saw 8 full garbage collections on startup (and, to be honest, a judicious 9th would've been good) with newrelic enabled, and without it there are none, but I haven't yet tested parallel=1 + newrelic=on, or Java11+newrelic GCs. Will try to do so on Monday, but I need to to get tools installed on the Java 11 image so I can compare, and... 🙄 . Please don't hold your breath, anyway.

**Comment by hughsimpson** [comment]
Have tested with `-Djava.util.concurrent.ForkJoinPool.common.parallelism=1` and my app is still seeing 8 FGCs on startup, even when that's the only difference between it and the nice, clean, no FGCs runs. Which is weird, and I don't really understand why. But I guess at this point we're well past the realms of newrelic-specific issues. I guess some work gets done on those pools that I usually manage without, but all a bit fishy-looking ... Nonetheless, as I say, limited to startup, and stable profile over time, so not a deal-breaker for me. I think the workaround should be officially advertised.

**Comment by ben-manes** [comment]
@hughsimpson pardon, but what is an FGC?

If I understand correctly, the summary is that some jdk17-based applications are being impacted by JDK-8274349. There are many consequences when to the `commonPool` is not functional, including the most obvious case resulting in a memory leak from NewRelic's caching. This is more of an early warning as many other problems may exist but be harder to diagnose. The resolution is for those users to either configure the JDK system property or ensure that they are on 17.0.2 or later (when released).

If that is the case then the fix for this issue is to warn users of the incompatibility, such as logging an error on startup if the incompatibility is detected.

Please confirm if my understanding is correct and @tbradellis if that suggested change would be an acceptable fix.

**Comment by Stephan202** [comment]
> @hughsimpson pardon, but what is an FGC?

IIUC it's short for "full heap garbage collection", usually called a "full GC". :)

Hugh, is the application startup actually slower, or could these full GCs (and associated extra memory pressure, I presume) be the result of otherwise-benign changes between JDK 11 and 17? I'm asking because I don't see how `-Djava.util.concurrent.ForkJoinPool.common.parallelism=1` can have a negative impact for applications impacted by JDK-8274349; after all, without the bug their parallelism would _also_ be 1. And since you also see the full GCs without New Relic, I'm just not sure what other variable there is :thinking:.

> If that is the case then the fix for this issue is to warn users of the incompatibility, such as by logging an error on startup if the incompatibility is detected.

I suppose that wouldn't hurt, but i.c.w. #572 I suppose New Relic customers wouldn't be impacted more by JDK-8274349 than any other JDK 17 user. (Which makes me wonder: do you plan to do "something" in Caffeine around this issue, since it impacts the default config for single-core JDK 17 users?)

**Comment by tbradellis** [comment]
> If that is the case then the fix for this issue is to warn users of the incompatibility, such as by logging an error on startup if the incompatibility is detected.
> 
> Please confirm if my understanding is correct and @tbradellis if that suggested change would be an acceptable fix.

Informing users of the incompatibility and pointing to JDK-8274349  does seem reasonable and would be in-line with how we've treated other issues that point back to the JDK (usually with some SIGSEGV on an older JDK). Although, if we can provide a code fix that avoids the issue we'd do that as well. We do plan to take https://github.com/newrelic/newrelic-java-agent/pull/572 as it's more in-line with the what was done in the old Guava dependency and if **I.U.C.** avoids this issue all together, and would therefore prevent time spent on future issue reports.

**Comment by ben-manes** [comment]
> We do plan to take #572 as it's more in-line with the what was done in the old Guava dependency and if I.U.C. avoids this issue all together, and would therefore prevent time spent on future issue reports.

I don't think it resolved the root problem because the `commonPool()` continued to not function. It did shift from being an easily detectable failure that could be traced to NewRelic, so the changes would avoid your team from being pointed at, but leave applications in a broken state. @Stephan202 is the circular initialization concern still valid or is JDK-8274349 the root problem?

> do you plan to do "something" in Caffeine around this issue, since it impacts the default config for single-core JDK 17 users?

@Stephan202 What do you think about having the builder check if `ForkJoinPool.getCommonPoolParallelism()` is less than 1 and, if so, defaulting to `Runnable::run`?

**Comment by Stephan202** [comment]
> @Stephan202 is the circular initialization concern still valid or is JDK-8274349 the root problem?

As best as I can tell the latter is the root cause, with New Relic not breaking initialization:
1. I just deployed the impacted application with New Relic Agent 7.4.0 but with a 2 CPU limit (rather than 1) and a heap dump shows that the common `ForkJoinPool` is properly initialized.
2. This also explains why several of our other Java 17 applications under much higher load aren't impacted.

> @Stephan202 What do you think about having the builder check if `ForkJoinPool.getCommonPoolParallelism()` is less than 1 and, if so, defaulting to `Runnable::run`?

I modified the [repro code](https://github.com/newrelic/newrelic-java-agent/pull/572#issuecomment-979441426) like this:
```diff
diff --git a/Test.java b/Test.java
index 03c8f97..b6e9f94 100644
--- a/Test.java
+++ b/Test.java
@@ -5,6 +5,7 @@ import java.util.concurrent.Semaphore;
 class Test {
   public static void main(String[] args) throws IllegalAccessException, InterruptedException {
     System.out.printf("Available processors: %s%n", Runtime.getRuntime().availableProcessors());
+    System.out.printf("Common FJP parallelism (1): %s%n", ForkJoinPool.getCommonPoolParallelism());
+    System.out.printf(
+        "Common FJP parallelism (2): %s%n", ForkJoinPool.commonPool().getParallelism());
 
     ForkJoinPool pool = ForkJoinPool.commonPool();
     printFields(System.out, pool);
```

But then the impacted cloud run still prints:
```
Available processors: 1
Common FJP parallelism (1): 1
Common FJP parallelism (2): 1
``` 
Part of this output is explained by the [output diff](https://github.com/newrelic/newrelic-java-agent/pull/572#issuecomment-979441426) shared earlier, as it shows that `COMMON_PARALLELISM` (which is used by the [first method](https://github.com/openjdk/jdk/blob/987992042454f92936d3efbd01e7beb921e3b70e/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2934-L2936)) is set to 1. IIUC the issue is restricted to the `SMASK` bits in `common.mode`. Unfortunately `ForkJoinPool.commonPool().getParallelism()` doesn't work either, because for some reason it "[hides](https://github.com/openjdk/jdk/blob/987992042454f92936d3efbd01e7beb921e3b70e/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2923-L2926)" the 0 case :thinking:. Right now I don't see an elegant solution yet.

NB: I wonder whether [this `Math.max`](https://github.com/openjdk/jdk/blob/987992042454f92936d3efbd01e7beb921e3b70e/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L3514) and [this conditional operator](https://github.com/openjdk/jdk/blob/987992042454f92936d3efbd01e7beb921e3b70e/src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java#L2925) should be there at all.

**Comment by Stephan202** [comment]
@ben-manes in case you want to reproduce the issue for yourself and play with potential identification mechanisms: I can confirm that `docker run --cpus 1` also triggers the bug.

**Comment by ben-manes** [comment]
I toyed with that docker flag quite a bit and I think checking the java version is unfortunately the best way. The snippet below is what I'm thinking of for 3.x and for 2.x it would have to check the `java.version` system property. wdyt?

```java
Executor getExecutor() {
  if ((executor != null)
      && !((executor instanceof ForkJoinPool) || (executor == ForkJoinPool.commonPool()))) {
    return executor;
  } else if ((ForkJoinPool.getCommonPoolParallelism() == 1) && (Runtime.version().feature() == 17)
      && (Runtime.version().interim() == 0) && (Runtime.version().update() <= 1)) {
    return Runnable::run; // JDK-8274349
  }
  return ForkJoinPool.commonPool();
}
```

**Comment by Stephan202** [comment]
The `instanceof` check is to avoid triggering `ForkJoinPool` class initialization? Cause it doesn't seem strictly necessary. If so: LGTM!

**Comment by ben-manes** [comment]
> The instanceof check is to avoid triggering ForkJoinPool class initialization? Cause it doesn't seem strictly necessary. If so: LGTM!

yep. It's probably not required and I could be convinced either way.

**Comment by Stephan202** [comment]
I suppose it won't hurt to keep. Attaching a debugger, I see that the pool isn't initialized some code explicitly triggers `ForkJoinPool` class initialization. So at the very least the extra conditional isn't vacuous from that POV.

**Comment by hughsimpson** [comment]
Yeah, sorry, FGC mean full garbage collection in this context. I am mystified by this and it's completely unrelated to newrelic. I suspect that there're some other environmental issues that're involved here from k8s or something, and it's probably not worth keeping any kind of running analysis on it in this thread. The important part is just that the unbounded memory growth stops when the common forkjoin pool has parallelism explicitly set, and the rest of what I've observed can probably be disregarded as weird project-specific artifacts

**Comment by ben-manes** [comment]
@Stephan202 When working on the release, I decided to remove the proposed detection and fallback logic. It would only mask the problem in some ways, but the application would be broken in any other usage. While it is very surprising that the commonPool is broken, it's a JVM issue that we can now explain to confused users. The only change that the library will make is to be more robust about evicting in this case, which would have avoided the memory leak.

I do think it might be reasonable for a java agent or similar tools to detect these types of bugs and warn users of problematic application settings. That fits with their goal of monitoring the system's health.

**Comment by Stephan202** [comment]
@ben-manes tnx for the update: that makes sense :)

As for the Java Agent warning: I suppose that could be turned into a full-fledged new feature: detection of in-use functionality impacted by known bugs (not just in the JDK, but also other libraries on the classpath). :chart_with_upwards_trend:

**Comment by ben-manes** [comment]
Caffeine's fix is released in v2.9.3 and v3.0.5. This will recover if the executor malfunctions, but affected users should apply a workaround for the JDK bug.

**Comment by mattnelson** [comment]
Was this confirmed to be impacting java 8 or 11?

**Comment by meiao** [comment]
@mattnelson this bug relates specifically to a but in Java 17, which will be fixed in Java 17.0.2.
We're updating the version of Caffeine on Java agent 7.5.0 which has a fix for this problem.

Check #623 for another possible cause of memory leak.

**Comment by hjain4** [comment]
Hi @ben-manes  we deployed new relic jdbc on production environment and looks code written by you popped up in thread dump which states thread in blocked due to new relic code acquiring a lock on resource. Can you please throw some light on it.

at java.util.regex.Pattern$CharPropertyGreedy.match(java.base@17.0.1.12.1/Unknown Source)
	at java.util.regex.Pattern$StartS.match(java.base@17.0.1.12.1/Unknown Source)
	at java.util.regex.Matcher.search(java.base@17.0.1.12.1/Unknown Source)
	at java.util.regex.Matcher.find(java.base@17.0.1.12.1/Unknown Source)
	at com.newrelic.agent.database.DefaultStatementFactory.parseStatement(DefaultStatementFactory.java:51)
	at com.newrelic.agent.database.DefaultDatabaseStatementParser.parseStatement(DefaultDatabaseStatementParser.java:106)
	at com.newrelic.agent.database.DefaultDatabaseStatementParser.getParsedDatabaseStatement(DefaultDatabaseStatementParser.java:99)
	at com.newrelic.agent.database.CachingDatabaseStatementParser.lambda$getParsedDatabaseStatement$0(CachingDatabaseStatementParser.java:64)
	at com.newrelic.agent.database.CachingDatabaseStatementParser$$Lambda$324/0x00000008006c54c0.apply(Unknown Source)
	at com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$doComputeIfAbsent$14(BoundedLocalCache.java:2406)
	at com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.BoundedLocalCache$$Lambda$237/0x0000000800491cf8.apply(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap.compute(java.base@17.0.1.12.1/Unknown Source)
	- locked <0x0000000598696c00> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
	at com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.BoundedLocalCache.doComputeIfAbsent(BoundedLocalCache.java:2404)
	at com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.BoundedLocalCache.computeIfAbsent(BoundedLocalCache.java:2387)
	at com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.LocalCache.computeIfAbsent(LocalCache.java:108)
	at com.newrelic.agent.deps.com.github.benmanes.caffeine.cache.LocalManualCache.get(LocalManualCache.java:62)
	at com.newrelic.agent.database.CachingDatabaseStatementParser.getParsedDatabaseStatement(CachingDatabaseStatementParser.java:62)
	at com.newrelic.agent.tracers.DefaultSqlTracer.parseStatement(DefaultSqlTracer.java:380)
	at com.newrelic.agent.tracers.DefaultSqlTracer.doFinish(DefaultSqlTracer.java:238)
	at com.newrelic.agent.tracers.DefaultTracer.performFinishWork(DefaultTracer.java:258)
	at com.newrelic.agent.tracers.DefaultTracer.finish(DefaultTracer.java:234)

This caused most of threads in blocked state due to 0x0000000598696c00 acquired by new relic code written by you

**Comment by ben-manes** [comment]
> This caused most of threads in blocked state due to 0x0000000598696c00 acquired by new relic code written by you

@hjain4 that stacktrace shows a computeIfAbsent-style call through a `ConcurrentHashMap`, which is blocked on a regex pattern search. From this alone there is nothing to indicate a problem with my code, so you would need to provide more information if you think that I can be of help. That snippet implies an expensive regex search. 

https://github.com/newrelic/newrelic-java-agent/blob/478ccc34989b05f7dce0333c7dc762d574cfa4cf/newrelic-agent/src/main/java/com/newrelic/agent/database/DefaultStatementFactory.java#L51
